be/src/exec/pipeline/exchange/local_exchange.cpp

```diff
@@ -174,6 +162,130 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
+            auto cmp = partition_columns[i]->compare_at(0, _previous_chunk->num_rows() - 1, *partition_columns[i], 1);
+            if (cmp != 0) {
+                is_joint_equal = false;
+                break;
+            }
+        }
+
+        if (!is_joint_equal) {
+            // The first row of current chunk is the start of a new partition, so
+            // send the chunk to the channel with the minimum number of rows.
+            channel_id = min_channel_id;
+        } else {
+            bool is_current_of_same_partition = true;
+            for (auto& column : partition_columns) {
+                auto cmp = column->compare_at(0, column->size() - 1, *column, 1);
+                if (cmp != 0) {
+                    is_current_of_same_partition = false;
+                    break;
+                }
+            }
+            if (is_current_of_same_partition) {
+                channel_id = _previous_channel_id;
+            } else {
+                // Found partition end that belongs to the first row of current chunk, and split the chunk into two parts:
+                // 1. The first part is the rows of the same partition as the last row of previous chunk, and send it to previous channel
+                // 2. The second part is the rows of the different partition, and send it to the channel with the minimum number of rows.
+
+                auto _find_first_not_equal = [](Column* column, int64_t target, int64_t start, int64_t end) {
+                    while (start + 1 < end) {
+                        int64_t mid = start + (end - start) / 2;
+                        if (column->compare_at(target, mid, *column, 1) == 0) {
+                            start = mid;
+                        } else {
+                            end = mid;
+                        }
+                    }
+                    if (column->compare_at(target, end - 1, *column, 1) == 0) {
+                        return end;
+                    }
+                    return end - 1;
+                };
+
+                int64_t end = chunk->num_rows();
+                for (auto& column : partition_columns) {
+                    end = _find_first_not_equal(column.get(), 0, 0, end);
+                }
+                // First part: [0, end)
+                ChunkPtr first_part = chunk->clone_empty();
+                first_part->append(*chunk, 0, end);
+                _source->get_sources()[_previous_channel_id]->add_chunk(first_part);
+
+                // Second part: [end, chunk->num_rows())
+                ChunkPtr second_part = chunk->clone_empty();
+                second_part->append(*chunk, end, chunk->num_rows() - end);
+                output_chunk = std::move(second_part);
+                channel_id = min_channel_id;
+            }
+        }
+    }
+
+    _source->get_sources()[channel_id]->add_chunk(output_chunk);
+    _previous_channel_id = channel_id;
+    _previous_chunk = chunk;
```

Comment:
it seems that we only  need old the last row of previous chunk, holding the whole previous chunk has a rick that PeakMemoryUsage grows.

Reply:
No big deal, only one chunk (4096 rows), and it will have extra overhead to keep only one last row.